File: python/linkedlist/sortList.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @return a ListNode
    def sortList(self, head):
        if head is None or head.next is None:
            return head

        fast = head
        slow = head
        while fast.next and fast.next.next:
            fast = fast.next.next
            slow = slow.next

        head1 = head
        head2 = slow.next
        slow.next = None

        head1 = self.sortList(head1)
        head2 = self.sortList(head2)
        head = self.merge(head1, head2)
        return head

    def merge(self, head1, head2):
        if head1 is None:
            return head2
        if head2 is None:
            return head1
        
        dummy = ListNode(0)
        p = dummy
        while head1 and head2:
            if head1.val < head2.val:
                p.next = head1
                head1 = head1.next
                p = p.next
            else:
                p.next = head2
                head2 = head2.next
                p = p.next

        if head1 is None:
            p.next = head2
        if head2 is None:
            p.next = head1

        return dummy.next

if __name__ == "__main__":
    solution = Solution()
    l0 = ListNode(8)
    l1 = ListNode(3)
    l2 = ListNode(7)
    l3 = ListNode(5)
    l4 = ListNode(9)
    l5 = ListNode(1)

    l0.next = l1
    l1.next = l2
    l2.next = l3
    l3.next = l4
    l4.next = l5

    p = solution.sortList(l0)
    while p is not None:
        print p.val
        p = p.next

File: java/SortList.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode sortList(ListNode head) {
        if ((head == null) || (head.next == null)) {
            return head;
        }
        ListNode slow = head;
        ListNode fast = head;
        while ((fast.next != null) && (fast.next.next != null)) {
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode f_half = head;
        ListNode s_half = slow.next;
        slow.next = null;
        ListNode head1 = this.sortList(f_half);
        ListNode head2 = this.sortList(s_half);
        ListNode new_head = this.merge(head1, head2);
        return new_head;
    }

    public ListNode merge(ListNode f_half, ListNode s_half) {
        if (f_half == null) {
            return s_half;
        }
        if (s_half == null) {
            return f_half;
        }
        ListNode dummy = new ListNode(0);
        ListNode p = dummy;
        while ((f_half != null) && (s_half != null)) {
            if (f_half.val > s_half.val) {
                p.next = s_half;
                s_half = s_half.next;
            } else {
                p.next = f_half;
                f_half = f_half.next;
            }
            p = p.next;
        }
        if (f_half != null) {
            p.next = f_half;
        }
        if (s_half != null) {
            p.next = s_half;
        }
        return dummy.next;
    }
}
